src/simperium/bucket.js

```diff
@@ -1,60 +1,435 @@
-	this.store = storeProvider( this );
-	this.isIndexing = false;
-}
+// @flow
+import events from 'events'
+import uuid from 'uuid/v4';
 
-inherits( Bucket, EventEmitter );
+const { EventEmitter } = events;
 
-Bucket.prototype.reload = function() {
-	this.emit( 'reload' );
-};
+type TaskCallback<T> = ( ?Error, ?T ) => void;
 
-Bucket.prototype.add = function( object, callback ) {
-	var id = uuid.v4();
-	return this.update( id, object, callback );
+/**
+ * Runs a promise with a callback (if one is provided) to support the old callback API.
+ * NOTE: if the callback API is removed this is a place to warn users
+ *
+ * @param {Function} [callback] - if provided, will be called with the expected values
+ * @param {Promise} promise - promise to run, executes callback if provieded
+ * @returns {Promise} promise is passed through
+ */
+const deprecateCallback = <T>( callback: ?TaskCallback<T>, promise: Promise<T> ): Promise<T> => {
+	if ( callback ) {
+		const perform = callback;
+		// Potentially could warn here if we decide to remove callback API
+		return promise.then(
+			( result ) => {
+				perform( null, result );
+				return result;
+			},
+			( error ) => {
+				perform( error, null );
+				return error;
+			}
+		);
+	}
+	return promise;
 };
 
-Bucket.prototype.get = function( id, callback ) {
-	return this.store.get( id, callback );
+export type BucketObject = {
+	id: string,
+	data: {},
+	isIndexing?: boolean
+}
+
+export interface BucketStore {
+	/**
+	 * Retrieve a bucket object from the store
+	 * @function
+	 * @name BucketStore#get
+	 * @param {String} id - the bucket object id to fetch
+	 * @param {bucketStoreGetCallback} - callback once the object is fetched
+	 */
+	get( id: string, callback: ( ?Error, ?BucketObject ) => void ): void;
+
+	/**
+	 * Updates the data for the given object id.
+	 *
+	 * @function
+	 * @name BucketStore#update
+	 * @param {String} id - to of object to update
+	 * @param {Object} data - data to update the object to
+	 * @param {Boolean} isIndexing - indicates the object is being downloaded during an index
+	 * @param {bucketStoreGetCallback}
+	 */
+	 update( id: string, data: {}, isIndexing: boolean, callback: ( ?Error, ?BucketObject ) => void ): void;
+
+	/**
+	 * Deletes the object at id from the datastore.
+	 *
+	 * @function
+	 * @name BucketStore#remove
+	 * @param {String} id - object to delete from the bucket
+	 * @param {bucketStoreRemoveCallback} - called once the object is deleted
+	 */
+	remove( id: string, callback: ( ?Error ) => void ): void;
+
+	/**
+	 * Fetchs all bucket objects from the datastore.
+	 *
+	 * @function
+	 * @name BucketStore#find
+	 * @param {?Object} query - currently undefined
+	 * @param {bucketStoreFindCallback} - called with results
+	 */
+	find( query: ?any, callback: ( ?Error, ?BucketObject[] ) => void ): void;
+}
+
+type BucketStoreProvider = ( Bucket ) => BucketStore;
+
+export type BucketObjectRevision = {
+	id: string,
+	version: number,
+	data: {}
 };
```

Comment:
sidenote: I find this extremely helpful. the hardest part in reading through this code or in the Simperium server's code isn't know when a type is invalid, but knowing when a collection of properties is supposed to be a given type, or knowing what properties to supply when a hidden remote function expects certain ones.

Reply:
I hope with fully defined `jsondiff` types this will shine a blazing late all expected properties in the diffing/syncing magic.